pkg/buildpack/buildpack.go

```diff
@@ -326,6 +356,133 @@ func ToLayerTar(dest string, module BuildModule) (string, error) {
 	return layerTar, nil
 }
 
+func ToNLayerTar(dest string, module BuildModule, logger logging.Logger) ([]ModuleTar, error) {
+	if !module.ContainsFlattenedModules() {
+		return handleSingleOrEmptyModule(dest, module)
+	}
+
+	modReader, err := module.Open()
+	if err != nil {
+		return nil, errors.Wrap(err, "opening blob")
+	}
+	defer modReader.Close()
+
+	tarCollection := newModuleTarCollection(dest)
+	rootFolderHeaders := map[string]*tar.Header{}
+
+	// the build module __could__ have more than one buildpack on it,
+	// if we detect multiple buildpacks, explode them
+	tr := tar.NewReader(modReader)
+	headersProcessed := 0
+	for {
+		header, err := tr.Next()
+		if err != nil {
+			if err == io.EOF {
+				if headersProcessed == 0 {
+					// If the reader was empty, we need to write an empty tar on disk
+					return handleSingleOrEmptyModule(dest, module)
+				}
+				break
+			}
+			return nil, errors.Wrapf(err, "failed to read next data '%s'", header.Name)
+		}
+
+		id, version := parseBpIDAndVersion(header)
+		if !matchIDVersion(module.Descriptor(), id, version) {
```

Comment:
Can we add a comment here about how this could happen?

Reply:
Suppose we have the following Build Module array = [ BM1, BM2 ] those are flattened so it means, when I open them and read their content I will find:
- 1 empty tarball 
- 1 tarball containing 
   -  /cnb/buildpacks/module-1/version-1 
   - /cnb/buildpacks/module-2/version-2

When we are iterating over the array [BM1, BM2].
**id, version** could be: (module-1, version-1) or (module-2,version-2)
**module** could be also BM1 or BM2. depending on which header we are reading from the tarball, those values could be the same or not, when they are not the same, it means *we are reading a flattened module from the tarball*

Reply:
Sounds good to me :) I think this extra bit of context will be helpful to future readers of this code.